### ISSUE TITLE
Brother HL-3140CW "size mismatch"

### ISSUE THREAD

**Issue by Mickkie** [apple/cups#5209]
A Brother HL-3140CW printer which was printing perfectly until recently, has started to print misaligned pages, with the top cropped by 13mm and the left side shifted to the left by 7mm.

The symptoms I've observed so far are as follows:

1.  Each time I modify a printer using the cups http GUI the 
corresponding ppd file default settings are messed up.  In particular, 
DefaultImageableArea and DefaultPaperDimension are reset to Letter for some 
reason:

grep Default /etc/cups/ppd/HL-3140CW.ppd
*DefaultColorSpace: Gray
*DefaultPageSize: A4
*DefaultPageRegion: A4
*DefaultImageableArea: Letter    <==
*DefaultPaperDimension: Letter  <==
*DefaultResolution: 300dpi

Selecting 'Set Default Options' in cups and (re)selecting A4 restores the 
correct settings in the ppd file:

grep Default /etc/cups/ppd/HL-3140CW.ppd
*DefaultColorSpace: Gray
*DefaultPageSize: A4
*DefaultPageRegion: A4
*DefaultImageableArea: A4
*DefaultPaperDimension: A4
*DefaultResolution: 300dpi
 
2. Setting the printer's Tray #1 from "Any" size to "A4" using the printer's http GUI results in the print job halted, as far as the printer is concerned, with the following message shown in its LED control panel:

 "Size mismatch. Load Letter paper and press Go."

The cups GUI reports the job as completed, although it never comes out of the printer - I cancel it at the printer manually.  I get this error whether I print a test page from cups/Maintenance/Print test page, or submit a print job of a text file using:

  lpr -o fit-to-page -o media=a4 -o page-ranges=1 -P 'HL-3140CW' test.txt

3. Setting the printer's Tray #1 from "A4" size back to "Any" results in misaligned print jobs: The top is cropped by ~13mm and the left side is shifted further to left by 7mm.

4. I tested the printer's page alignment by printing a test page using its own panel and by using MSWindows 7 running in a VM.  The pages are aligned properly, just as they used to be printing from Linux in the recent past.  The MSWindows jobs are printed aligned whether I have set the pinter's Tray #1 to size "Any" or "A4".

5. The problem started while I was using cups-2.1.4 and cups-filters-1.16.4 on Gentoo, but I have since recompiled these with gcc-6.4.0 to versions cups-2.2.6 and cups-filters-1.18.0 respectively, with no effect.  I have not been able to identify which other package update might have had an effect on this problem (e.g. poppler?).

6. The Brother's driver and cups-wrapper installed are:

http://support.brother.com/g/b/files/dlf/dlf007067/hl3140cwlpr-1.1.2-1.i386.rpm
http://download.brother.com/welcome/dlf007069/hl3140cwcupswrapper-1.1.4-0.i386.rpm

I've looked at the cups debug output but couldn't spot anything amiss as far as page size is concerned, although pstops reports it can't find a resolution specified in ppd:

D [28/Dec/2017:11:39:00 +0000] [Job 668] Running command line for pstops: pstops 668 anonymous \'Test Page\' 1 \'job-uuid=urn:uuid:f5563200-20b3-3279-600d-0bfab0eaaa1e job-originating-host-name=localhost date-time-at-creation= date-time-at-processing= time-at-creation=1514461140 time-at-processing=1514461140\'
D [28/Dec/2017:11:39:00 +0000] [Job 668] No resolution information found in the PPD file.
D [28/Dec/2017:11:39:00 +0000] [Job 668] Using image rendering resolution 300 dpi
D [28/Dec/2017:11:39:00 +0000] [Job 668] Using image rendering resolution 300 dpi

Which is odd, because ppd specifies:

  *DefaultBRResolution: 600dpi

Then pdftops seems to handle the printing of the test page:

D [28/Dec/2017:11:39:00 +0000] [Job 668] Running command line for pdftops: pdftops -level3 -origpagesizes -nocenter -r 300 /var/spool/cups/tmp/053875a5282ee -
D [28/Dec/2017:11:39:00 +0000] [Job 668] Started filter pdftops (PID 21395)
D [28/Dec/2017:11:39:00 +0000] [Job 668] Started filter pstops (PID 21396)
D [28/Dec/2017:11:39:00 +0000] [Job 668] Page = 595x842; 12,12 to 583,830
D [28/Dec/2017:11:39:00 +0000] [Job 668] slow_collate=0, slow_duplex=0, slow_order=0
D [28/Dec/2017:11:39:00 +0000] [Job 668] Before copy_comments - %!PS-Adobe-3.0
D [28/Dec/2017:11:39:00 +0000] [Job 668] %!PS-Adobe-3.0
D [28/Dec/2017:11:39:00 +0000] [Job 668] %Produced by poppler pdftops version: 0.57.0 (http://poppler.freedesktop.org)
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%Creator: cairo 1.10.2 (http://cairographics.org)
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%LanguageLevel: 3
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%DocumentSuppliedResources: (atend)
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%DocumentMedia: A4 595 842 0 () ()
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%BoundingBox: 0 0 595 842
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%Pages: 1
D [28/Dec/2017:11:39:00 +0000] [Job 668] %%EndComments
D [28/Dec/2017:11:39:00 +0000] [Job 668] Before copy_prolog - %%BeginProlog
D [28/Dec/2017:11:39:00 +0000] [Job 668] Before copy_setup - %%BeginSetup
D [28/Dec/2017:11:39:01 +0000] [Job 668] Before page loop - %%Page: 1 1
D [28/Dec/2017:11:39:01 +0000] [Job 668] Copying page 1...
D [28/Dec/2017:11:39:01 +0000] [Job 668] pagew = 571.0, pagel = 818.0
D [28/Dec/2017:11:39:01 +0000] [Job 668] bboxx = 0, bboxy = 0, bboxw = 595, bboxl = 842
D [28/Dec/2017:11:39:01 +0000] [Job 668] PageLeft = 12.0, PageRight = 583.0
D [28/Dec/2017:11:39:01 +0000] [Job 668] PageTop = 830.0, PageBottom = 12.0
D [28/Dec/2017:11:39:01 +0000] [Job 668] PageWidth = 595.0, PageLength = 842.0
D [28/Dec/2017:11:39:01 +0000] [Job 668] Wrote 1 pages...
D [28/Dec/2017:11:39:01 +0000] [Job 668] PID 21395 (pdftops) exited with no errors.
D [28/Dec/2017:11:39:01 +0000] [Job 668] PID 21396 (pstops) exited with no errors.
D [28/Dec/2017:11:39:01 +0000] [Job 668] PID 21383 (/usr/libexec/cups/filter/pdftops) exited with no errors.
D [28/Dec/2017:11:39:01 +0000] [Job 668] rm: cannot remove \'/var/tmp/lprng_hl3140cw_rcname\': Operation not permitted
D [28/Dec/2017:11:39:02 +0000] [Job 668] Copying print data.
D [28/Dec/2017:11:39:02 +0000] cupsdMarkDirty(---J-)

Please ask if you need more information.
-- 
Regards,
Mick

**Comment by Mickkie** [comment]
I found a workaround, although this does not explain why cups cannot set the PageSize in ppd fully when modifying a printer and why the ppd PageSize setting is ignored in any case.  I noticed that brother installs two files:

 /opt/brother/Printers/hl3140cw/inf/brhl3140cwfunc

where [Default] settings are specified and

 /opt/brother/Printers/hl3140cw/inf/brhl3140cwrc

where printer specific [hl3140cw] settings are defined.  I changed the setting to "PageSize=A4" in the second file and now pages are printed in A4 as they always used to.  This does not address the underlying problem of cups not being able to set a default page size in a way that is respected by the printer, so a more intelligent fix is probably necessary.
-- 
Regards,
Mick
